src/libutil/source-accessor.hh

```diff
@@ -137,6 +137,15 @@ struct SourceAccessor
     void setPathDisplay(std::string displayPrefix, std::string displaySuffix = "");
 
     virtual std::string showPath(const CanonPath & path);
+
+    /**
+     * System paths: `toString /foo/bar = "/foo/bar"`
+     * Virtual paths: fetched to the store
+     *
+     * In both cases, the returned string functionally identifies the path,
+     * and can still be read.
+     */
+    virtual bool toStringReturnsStorePath() const;
```

Comment:
The only truly new semantics.

Either way, both kinds of paths are virtual in the sense that they haven't been copied yet.
If you have copied it, it'd be a string. (Not _if and only if_, although Nix does discourage that.)

In lazy-trees, this could be changed to a virtual store path string without a problem.

Reply:
TODO: forward port this rule to prove it.

Reply:
https://github.com/NixOS/nix/pull/10511#issuecomment-2057564588 I think rather contain information with which to construct the path

Reply:
Btw, here I've picked the least breaking semantics:

- The return value still uniquely identifies the source
- Still deterministic
- Still readable when converted to a system path with `/. + x` or `/${x}`

Performance is at least as good as the status quo, but not as good as lazy trees, because for `toString x` to be instant, you need to sacrifice one of the above, or come up with a clever scheme, like opaque placeholders in strings or something.